lib/datadog/grok/src/parse_grok_rules.rs

```diff
@@ -0,0 +1,375 @@
+use std::{collections::HashMap, convert::TryFrom, fmt::Write, sync::Arc};
+
+use grok::Grok;
+use lazy_static::lazy_static;
+use regex::Regex;
+
+use lookup::LookupBuf;
+use parsing::value::Value;
+
+use crate::ast::{self, Destination, GrokPattern};
+use crate::grok_filter::GrokFilter;
+use crate::parse_grok_pattern::parse_grok_pattern;
+use itertools::{Itertools, Position};
+
+#[derive(Debug, Clone)]
+pub struct GrokRule {
+    pub pattern: Arc<grok::Pattern>,
+    pub filters: HashMap<LookupBuf, Vec<GrokFilter>>,
+}
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum Error {
+    #[error("failed to parse grok expression '{}': {}", .0, .1)]
+    InvalidGrokExpression(String, String),
+    #[error("invalid arguments for the function '{}'", .0)]
+    InvalidFunctionArguments(String),
+    #[error("unknown filter '{}'", .0)]
+    UnknownFilter(String),
+}
+
+/**
+Parses DD grok rules.
+
+Here is an example:
+access.common %{_client_ip} %{_ident} %{_auth} \[%{_date_access}\] "(?>%{_method} |)%{_url}(?> %{_version}|)" %{_status_code} (?>%{_bytes_written}|-)
+access.combined %{access.common} (%{number:duration:scale(1000000000)} )?"%{_referer}" "%{_user_agent}"( "%{_x_forwarded_for}")?.*"#
+
+You can write parsing rules with the %{MATCHER:EXTRACT:FILTER} syntax:
+- Matcher: A rule (possibly a reference to another token rule) that describes what to expect (number, word, notSpace, etc.)
+- Extract (optional): An identifier representing the capture destination for the piece of text matched by the Matcher.
+- Filter (optional): A post-processor of the match to transform it.
+
+Each rule can reference parsing rules defined above itself in the list.
+Only one can match any given log. The first one that matches, from top to bottom, is the one that does the parsing.
+For further documentation and the full list of available matcher and filters check out https://docs.datadoghq.com/logs/processing/parsing
+*/
+pub fn parse_grok_rules(
+    helper_rules: &[String],
+    parsing_rules: &[String],
+) -> Result<Vec<GrokRule>, Error> {
+    let mut prev_rule_definitions: HashMap<&str, String> = HashMap::new();
+    let mut prev_rule_filters: HashMap<&str, HashMap<LookupBuf, Vec<GrokFilter>>> = HashMap::new();
+    let mut grok = initialize_grok();
+
+    // parse helper rules to reference them later in the match rules
+    parse_rules(
+        helper_rules,
+        &mut prev_rule_definitions,
+        &mut prev_rule_filters,
+        &mut grok,
+    )?;
+    // parse match rules and return them
+    parse_rules(
+        parsing_rules,
+        &mut prev_rule_definitions,
+        &mut prev_rule_filters,
+        &mut grok,
+    )
+}
+
+fn parse_rules<'a>(
```

Comment:
It would be useful to document what the parameters `prev_rule_definitions` and `prev_rule_filters` are used for. It's not instantly apparent to me.

Reply:
Yeah, definitely. Refactored and added some comments in https://github.com/timberio/vector/pull/8850/commits/5f3fc1072833e91d05199a642354d3c52962fe9d